oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,122 @@
+
+Alternatives Considered
+-----------------------
+
+We chose to avoid putting the locale in the domain name itself, to leave that available for other uses. We considered but rejected putting it in as a URL parameter; that can be harder to parse out and doesn't give us automatic separation of analytics by locale. Using the first component of the path ensures that if users send a link or bookmark a page, their recipients will see the exact content that they intended to share.
+
+For the locale selection strategy, we considered but rejected a fallback step where we would choose a language based on the geographic location of the user's IP address. This is generally not reliable: many countries have speakers of multiple languages, users might be traveling, and users might be living or studying in a country but still be more comfortable in a different language.
+
+Accessibility Requirements
+==========================
+
+Translate all non-display text that is present in a tag for accessibility reasons (e.g., ``aria-label``), using the same locale as the rest of the page.
+
+There is a special case where a message id doesn't have a translation in the user's preferred locale, so we fall back to a default locale. In this case, that string needs to be wrapped in a ``<span locale="....">`` tag to maintain the correct locale information for accessibility tools. In order to make sure that this string wrapping always happens, edX has written ``FMessage``, a simple wrapper around the standard react-intl ``FormattedMessage`` component, which will be the standard component used to format strings. ``FMessage`` can be found in the **[TODO]** repo.
+
+Message Ids
+===========
+
+Message ids, or message keys, uniquely identify a string to be internationalized. They should consist of one or more words to scope the message, plus one or more words to identify the message. All words will be lowercase and separated by periods. Here are examples of message ids used in a header and in a login modal, respectively:
+
+  | ``header.search``
+  | ``login.no.such.password``
+  | ``login.privacy.policy``
+
+The part of the id used for scoping is purely to reduce collisions and help human readability; it's not intended to be programmatically parsable.
+
+Alternatives Considered
+-----------------------
+
+Previous i18n implementations have sometimes used the English text as the message id. While this saves developers a step in their work, it has the drawback that a change in the English copy invalidates all existing translations, even if they would not otherwise have needed to change. (For example, capitalizing a noun in English has no impact on the German translation, because nouns are capitalized in German anyway, or an English change might be correcting a typo that was correct in the translations.) Do not use English copy as the message id.
+
+Dates And Numbers
+=================
+
+All dates and numbers should be localized.
```

Comment:
It may be considered unclear whether this applies to phone numbers in addition to counting numbers. Is it unacceptable scope creep to suggest that we should always give phone numbers in the format of `+<country code> <number with spaces>` without the American-traditional dashes and parentheses?

Reply:
Good question!  I'm not sure we have a decision right now, but if one comes up, we will add it in.